### ISSUE TITLE
E2EE: Messages sent from a new E2EE room are unreadable by other clients (Riot)

### ISSUE THREAD

**Issue by lbeltrame** [mujx/nheko#341]
### System:

- Nheko version: 0.43-8704265
- Installation method: Source
- Operating System: openSUSE Tumbleweed
- Qt version: 5.11
- C++ compiler: GCC 8
- Desktop Environment:KDE Plasma 5.13.80 (git master version)
- Self-hosted Synapse homeserver (0.31.2)
- Electron Riot client

### Actual behavior

When creating a new room in either nheko or Riot and then enabling E2EE (it doesn't matter which), messages sent from nheko will not be decrypted on the Riot side with this output:

```
** Unable to decrypt: Unknown encryption algorithm "". **
```

Messages sent from Riot are however correctly handled by nheko.

The raw source for a message just containing the word "test" is as follows:

```
{
  "origin_server_ts": TIMESTAMP,
  "sender": "@my_user_name:example.com",
  "event_id": "EVENT_ID",
  "unsigned": {
    "age": 26
  },
  "content": {
    "sender_key": "SENDER_KEY",
    "ciphertext": "CRYPTED_CONTENT ('test')",
    "session_id": "SESSION_ID",
    "algorithm": "",
    "device_id": "MY_DEVICE_ID"
  },
  "type": "m.room.encrypted",
  "room_id": "MY_ROOM_ID"
}
```

On the nheko side (log output) the message is correctly decrypted:

```
[2018-06-18 14:25:38.272] [crypto] [info] next message_index 2
[2018-06-18 14:25:38.331] [ui] [info] [XXXX] sending encrypted event
[2018-06-18 14:25:38.547] [ui] [info] [XXXX] message was received by the server
[2018-06-18 14:25:38.765] [crypto] [info] decrypted data: 
 {
  "content": {
    "body": "test",
    "msgtype": "m.text"
  },
  "event_id": "EVENT_ID",
  "origin_server_ts": TIMESTAMP,
  "room_id": "MY_ROOM_ID",
  "sender": "@einar:conference.heavensinferno.net",
  "type": "m.room.message",
  "unsigned": {
    "age": 212,
    "transaction_id": "XXXX"
  }
}
```

Things like transaction ID and timestamps correspond.

### Expected behavior

Messages should be properly decrypted when they reach Riot.

### Steps to reproduce

1. Launch nheko in a terminal
2. Type or receive an event in an encrypted room
3. Watch nheko's standard output

**Comment by mujx** [comment]
@lbeltrame Could you try with the current master (604cdcec8abf8a8c564111d6f55feaf94486ba6c) ? Does this still happen if you create a new device (logout & re-login with nheko) ?

Also it would be helpful if you could paste a bigger part of the logs (i.e when sending and receiving messages).

**Comment by lbeltrame** [comment]
I'll test shortly and report back.

**Comment by lbeltrame** [comment]
Apparently there's no change (updated, wiped the build dir, rebuilt, logged out/in).

Logs:

```
[2018-06-18 14:47:18.981] [crypto] [info] established inbound megolm session (XXXX)
[2018-06-18 14:47:19.352] [crypto] [info] decrypted event: XXXX
[2018-06-18 14:47:26.440] [crypto] [info] received 1 to_device messages
[2018-06-18 14:47:26.440] [crypto] [info] sender    : @me:example_com
[2018-06-18 14:47:26.440] [crypto] [info] sender_key: KEY
[2018-06-18 14:47:26.440] [crypto] [info] type: OLM_MESSAGE
[2018-06-18 14:47:26.440] [crypto] [info] attempt to decrypt message with 1 known session_ids
[2018-06-18 14:47:26.473] [crypto] [info] decrypted olm payload: {
  "content": {
    "algorithm": "m.megolm.v1.aes-sha2",
    "chain_index": 0,
    "room_id": "ROOM_ID",
    "session_id": "SESSION_ID",
    "session_key": "SESSION_KEY"
  },
  "keys": {
    "ed25519": "KEY"
  },
  "recipient": "@me:example.com",
  "recipient_keys": {
    "ed25519": "KEY"
  },
  "sender": "@me:example.com",
  "sender_device": "DEVICE_ID",
  "type": "m.room_key"
}
[2018-06-18 14:47:26.515] [crypto] [info] established inbound megolm session (XXXX)
# When sending the message
[2018-06-18 14:47:39.340] [ui] [info] [g2SNTSoiIrxJJupRDQ2e6cqHaBtH6n46] sending next queued message
[2018-06-18 14:47:39.340] [ui] [info] [g2SNTSoiIrxJJupRDQ2e6cqHaBtH6n46] sending encrypted event
[2018-06-18 14:47:39.341] [crypto] [info] next message_index 2
[2018-06-18 14:47:39.617] [ui] [info] [g2SNTSoiIrxJJupRDQ2e6cqHaBtH6n46] message was received by the server
[2018-06-18 14:47:39.815] [crypto] [info] decrypted event: XXXX
```

**Comment by Ezwen** [comment]
I tried e2e for the first time after building `master`, and I could not reproduce this problem. e2e seems to be working quite well.

**Comment by lbeltrame** [comment]
Would logs of the message exchange from Synapse help, or is this a purely client side matter?

**Comment by lbeltrame** [comment]
Unable to reproduce with the latest master. Probably 7ecabcd61410783267f1ef4353d65716b84f7f83 fixed this as well. Closing.

**Comment by lbeltrame** [comment]
I updated to the latest master state of matrix-structs, mtxclient, and nheko, and apparently this problem resurfaced. I can't see what changed though.

**Comment by lbeltrame** [comment]
Even after a clean build and clean cache, the problem persists. It has somehow regressed.

**Comment by lbeltrame** [comment]
My previously-built version keeps on working, so I guess there was a regression somewhere else in the dependencies.

**Comment by mujx** [comment]
@lbeltrame Could you send those two version so we can find out what changed?

**Comment by lbeltrame** [comment]
Non-working version is:

nheko - cb21433
matrix-structs - eeb7373
mtxclient - 96fd35e

Working version is

nheko - 74396e9
matrix-structs - c115757
mtxclient - d5ee30

**Comment by mujx** [comment]
There is nothing in between those commits that could cause a regression (most of them are packaging related).

Could you elaborate in more detail on the timeline of events? Also include some logs between the events to help us better understand the problem.

What do you mean by `the previous build keeps working`? Do you use the same device/login credentials but with another instance of nheko, or logout/login to a new device?

**Comment by lbeltrame** [comment]
The timeline of events is as follows:

From the "working" version (tested):

- Update all the dependencies
- Build and install the dependencies
- Build nheko

Afterwards people reported that they were no longer able to see my messages. I'll send the logs later today when I turn that machine back on.

**Comment by mujx** [comment]
So the working & non working versions were on different machines?

**Comment by lbeltrame** [comment]
Sorry, I explained myself *very* badly.

The version on the different machine is one that I purposefully kept back. 

The revisions I listed as "working" were also on the machine I saw this issue (tested and re-tested). After updating there, I saw this (possible) regression. I held back everything else until I was sure it was not just a local issue (but also wiping caches etc won't help).

I will roll back the "affected" machine to the previous revisions to make sure there isn't something else (but I'll still post those logs!).
